### PR TITLE
Query Engine

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request refactors the query engine from using a two-level map/reduce system to an n-level iterator-based system. The previous system had a number of issues that warranted a refactor:

1. Map/reduce limits the types of queries that can be done (e.g. nested queries)

2. Used a cursor-based system with very loose types which caused excessive memory allocation because it needed to type convert to `interface{}`.

3. Was very complicated after several iterations of multiple developers adding numerous features.

This refactor addresses these issues:

1. Uses n-level nested iterators to stream, merge, and join point data. This allows more complicated queries such as nested queries and also allows aggregate functions to be developed more easily.

2. All points have a specific data type and iterators can use buffering and flyweights to reduce or eliminate memory allocations altogether.

3. Simplifies the query system by making everything an "iterator" with well defined levels of execution.

## Iterators

The core concept in the new engine is the `influxql.Iterator`. There are separate iterator interfaces for each type (`Float`, `Integer`, `String`, `Boolean`) and each one produces a stream of typed points (`FloatPoint`, `IntegerPoint`, `StringPoint`, `BooleanPoint`).

When a `*SelectStatement` comes into the system, it is passed to `influxql.Select()` which generates a slice of iterators -- one per field. For each field, the engine generates an `IteratorOptions` to hold the expression, conditions, time range, etc and passes that to the `IteratorCreator` (which at the top level is implemented by the `tsdb.Store`).

The `IteratorCreator` is an interface that takes an `IteratorOptions` and returns an `Iterator`. The store will pass off to the shard and the shard will pass off to the engine. Each of these levels is an `IteratorCreator` and handles creating child iterators and appropriately merging them together. The final level (the engine) receives the options and has the ability to decide how to implement it. Most of the time it will take an expression such as `sum(value)` and create an iterator for the `value` field and wrap it in a `SumIterator`, however, optimizations can be added such as returning a precomputed `count()` instead of iterating over all the values.

Once the iterators are returned from the iterator they are joined (via `influxql.Join()`) and passed to the `influxql.Emitter` which emits `model.Row` objects with appropriate grouping of name and tags.

## TODO

This pull request is a work in progress, however, it is large so I wanted to get it pushed up for review as soon as possible. The following are a list of known issues that still need to be resolved. They are all relatively minor.

- [x] `LIMIT`/`OFFSET`
- [x] `SLIMIT`/`SOFFSET`
- [x] `Integer` type
- [x] Convert `derivative()` to iterator form
- [x] Convert `top()`/`bottom()` to iterator form
- [x] Binary expression evaluation
- [x] Fills
- [x] Convert `SHOW` commands to use iterators

### CODE REVIEW

**Comment by otoolep** [comment]
Unfortunately github won't show us the entire diff. :-(

**Comment by benbjohnson** [comment]
@otoolep Diffs are for the weak. :)

**Comment by otoolep** [comment]
Yeah, I can that! Time for file-level review, methinks.

**Comment by pauldix** [comment]
Looks like the cluster (remote iterator) stuff also still needs to be wired up?

File: influxql/iterator.gen.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+	min := ZeroTime
+	for _, itr := range a {
+		// Read next point. Ignore if no more points available.
+		p := itr.Next()
+		if p == nil {
+			continue
+		}
+
+		// Update values if point's window is lower (sorted by <name,tags,time>).
+		if min == ZeroTime || p.Name < name || (p.Name == name && p.Tags.ID() < tags.ID()) || (p.Name == name && p.Tags.Equals(&tags) && p.Time < min) {
```
Comment (nathanielc) [inline influxql/iterator.gen.go.tmpl]: What is a Point's name and why does its sort order matter?

**Comment by benbjohnson** [comment]
@pauldix Yes, I was going off my list based on the test suite and I missed that. I added it to the PR's TODO.

File: influxql/iterator.gen.go.tmpl (hunk 1 above)
Comment (benbjohnson) [inline influxql/iterator.gen.go.tmpl]: It's the measurement name. Points get grouped by name and tags by the emitter so they get sorted together too.

Comment (nathanielc) [inline influxql/iterator.gen.go.tmpl]: OK, so not only can an underlying iterator produce results for different groups but data from different measurements too, makes sense.

**Comment by nathanielc** [comment]
So one thing I haven't been able to quite understand yet is the `map` phase for the map/reduce iterators.

For many map/reduce iterator the map and reduce steps are the same; `count` for example,
but for others it is not i.e. `mean`. I see that there are ReduceIterators and ReduceSliceIterators and it seems like the functions that don't really need a map phase just use the ReduceIterator interface. The ReduceIterator step is applied as many times are necessary to arrive at a final result. The functions that do need a map phase use the ReduceSliceIterator. Does a ReduceSliceIterator guarantee that all needed values are present in the slice or is the ReduceSlice step applied multiple times too? In other words, will the output of a ReduceSliceIterator every be passed back as input to the next iteration of a ReduceSliceIterator? 

If it is guaranteed that all needed values are present in the slice how and why? Meaning why not have an explicit map phase so that only intermediate results need be passed to the reduce phase. Is this an artifact of the storage layer where all needed values are guaranteed to live in the same shard etc.

**Comment by benbjohnson** [comment]
@nathanielc The `ReduceSliceIterator` guarantees that all points in a group (name + tags + window) are present before the aggregate is computed. This is required for a few aggregate functions (e.g. `median()`, `stddev()`, & `percentile()`).

Some of the aggregate functions such as `mean()` & `distinct()` are computed using `ReduceSliceIterator` right now because it's easier. They actually need their own specific iterator types to implement an optimized map/reduce. It's not a difficult change to make but I wanted to get all the tests passing before I move on to optimizations.

**Comment by benbjohnson** [comment]
@gunnaraasen I just pushed up a working copy of the code: c9a2eb3.

**Comment by nathanielc** [comment]
@benbjohnson That makes sense, so taking mean for example to optimize it would you create a new point type called MeanPoint or something that contains both the sum and count of a subset of the group, and then MeanPoint would be reduced into the final FloatPoint result of the mean? Or would you not bother with a new type and somehow make it fit into the existing types?

**Comment by li-ang** [comment]
@benbjohnson Could you tell me when will you finish the PR?  Thanks and Happy New Year!

**Comment by benbjohnson** [comment]
@jsternberg I went through the test suite again to see where we're at and what needs to be done. I compiled the full list here:

https://gist.github.com/benbjohnson/02c8504e7c5e307f81c8

**Comment by benbjohnson** [comment]
Query engine is complete (except for remote execution). Please review. Remote exec will be done in a separate PR.

**Comment by jwilder** [comment]
Circle is failing due to `go vet` errors:

```
Go vet failed. Please run 'go vet ./...' and fix any errors.

tsdb/query_executor.go:567: influxql.Fields composite literal uses unkeyed fields
tsdb/query_executor.go:570: influxql.Sources composite literal uses unkeyed fields
tsdb/query_executor.go:624: influxql.Fields composite literal uses unkeyed fields
```

**Comment by e-dard** [comment]
Looked through PR. Looks good from what I can see. Couple of small typo-type nits:

- iterators.gen.go.tml:129: `anymore`
- call_iterator.go:25: `auxilary`
- iterator.go:556: `Series` could have a comment (also the `ID` and `Combine` methods)
- iterator.go:578: `SeriesList` could have a comment.

**Comment by e-dard** [comment]
I think someone mentioned that `SHOW TAG KEYS` should be implemented for a cluster, in this branch, but local testing of this branch indicates that it doesn't fix #5591. This is a note to myself, more than anything :smile:

**Comment by jsternberg** [comment]
Auxiliary is misspelled in most comments. I'll make a follow up commit after this gets merged for those, but since they're comments, I think they're immaterial for initial approval so others can start building on this.

Anymore is a US spelling which is probably why you're not familiar with it (http://alt-usage-english.org/anymore.html).

I'll add documentation comments to the series stuff. I might remove those later since it was primarily used for an implementation of fill that doesn't exist anymore. Now it is used for determining the types of auxiliary iterators, but there's a much more efficient implementation that should probably be used instead.

Any cluster commands won't work until we have remote exec which isn't part of the initial PR.

**Comment by jwilder** [comment]
:+1:

**Comment by seiflotfy** [comment]
SWEEEEEET!

File: cmd/influxd/run/select_test.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package run_test
```
Comment (li-ang) [inline cmd/influxd/run/select_test.go]: Why no test cases?

Comment (li-ang) [inline cmd/influxd/run/select_test.go]: `influxql/select_test.go` is the right test file ?

Comment (jsternberg) [inline cmd/influxd/run/select_test.go]: Likely just a mistake. The tests are in `cmd/influxd/run/server_test.go`.

Comment (li-ang) [inline cmd/influxd/run/select_test.go]: https://github.com/influxdata/influxdb/pull/5622

**Pull request merged**